### PR TITLE
Infer step size for Embeddings

### PR DESCRIPTION
**Pull request by KuuCi**
Theres two cases: 

1. For in batch negatives, total number of negatives is batch size - 1. And pos_step_size is 2. 
2. For hard negatives, each sample has a list of negatives that can be as long as you like. In this case you have to change the pos_step_size.

We can remove setting step_size and autoinfer without needing the pos_step_size. We will need to look at the first batch to determine how many hard negatives. If there are hard negatives, we will use that - 1 as the step size. Otherwise, we will defer to batch negatives, setting step_size = 2

Testing:

embedding-ft-no-pos-zXOtvz (off of this branch, removed pos_step_size keyword):
image

embedding-ft-pos-kl2fPy (off foundry main pos_step_size = 2):
image

embedding-ft-pos-neg-gMJSGx (off foundry main pos_step_size = 21)
image

embedding-ft-no-pos-neg-3GoEms (off of this branch, removed pos_step_size keyword):
image

### CODE REVIEW

File: llmfoundry/models/llm_embed/modeling_llm_embed.py [hunk 1]
```diff
@@ -208,6 +207,19 @@ def construct_model(self):
             model = MPTForCausalLM(MPTConfig(**self.kwargs))
             self.is_mpt = True
         return model
+    
+    def _update_step_size_if_needed(self, batch: MutableMapping) -> None:
+        """Update step size on first batch if we detect hard negatives"""
+        if self._first_batch_seen:
+            return
+
+        input_shape = batch['input_ids'].shape
```
[the remark this answers is not in the document]
Comment (jacobfulano) [reply llmfoundry/models/llm_embed/modeling_llm_embed.py:216]: if there is 1 hard negative:
input_shape[1] = 3 ?
if there are 2 hard negatives:
input_shape[1] = 4?

If so, then wouldn't we want `self.step_size = input_shape[1]`? Please correct me if I'm wrong

[the remark this answers is not in the document]
Comment (KuuCi) [reply llmfoundry/models/llm_embed/modeling_llm_embed.py:216]: I am pretty positive that query is not included in the batch, just [positive + negative]

[the remark this answers is not in the document]
Comment (KuuCi) [reply llmfoundry/models/llm_embed/modeling_llm_embed.py:216]: I think this is correct, updating code.

**Comment by mrdrozdov** [comment]
I really like this idea. I do have a couple soft suggestions:

1. Rather than `self._first_batch_seen` can you simply set `self.step_size = None` during init? Then check `if self.step_size is None` instead of using first_batch_seen.

2. It could be more tidy to just make step_size part of the batch. I think you could do something like `batch["step_size"] = step_size`, so that it is accessible in compute_score? That being said, I'm not familiar enough to foundry to know if this causes complications, and you might need to do something like `batch["step_size"] = torch.tensor([step_size] * batch_size, dtype=torch.long)`. If you go this route, then you can ignore (1).

Anyway, the current approach is fine too, but wanted to share just in case.

File: llmfoundry/models/llm_embed/modeling_llm_embed.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (jacobfulano) [reply llmfoundry/models/llm_embed/modeling_llm_embed.py:215]: I'm with @mrdrozdov here. 

The separate columns in query, passage, negatives are flattened into a single list here (e.g. [Q,P,N,Q,P,N]): https://github.com/mosaicml/llm-foundry/blob/bec9ecdbee9965f507b479d39cf5465ddfc75bc4/llmfoundry/data/contrastive_pairs/dataloader.py#L172

This is the code that selects then each query after it has been tokenized [Q,P,N,Q,P,N] --> [Q,Q]: https://github.com/mosaicml/llm-foundry/blob/bec9ecdbee9965f507b479d39cf5465ddfc75bc4/llmfoundry/models/llm_embed/modeling_llm_embed.py#L212

[the remark this answers is not in the document]
Comment (jacobfulano) [reply llmfoundry/models/llm_embed/modeling_llm_embed.py:215]: Once we confirm this, we should probably update the comments here: https://github.com/mosaicml/llm-foundry/blob/bec9ecdbee9965f507b479d39cf5465ddfc75bc4/llmfoundry/models/llm_embed/modeling_llm_embed.py#L219 and here https://github.com/mosaicml/llm-foundry/blob/bec9ecdbee9965f507b479d39cf5465ddfc75bc4/llmfoundry/models/llm_embed/modeling_llm_embed.py#L237

[the remark this answers is not in the document]
Comment (KuuCi) [reply llmfoundry/models/llm_embed/modeling_llm_embed.py:215]: def feels strange that that manual tests yielded same results for when we has `pos_step_size` and then using this pr to infer it

[the remark this answers is not in the document]
Comment (KuuCi) [reply llmfoundry/models/llm_embed/modeling_llm_embed.py:215]: image
This is what happens when `self.step_size = input_shape[1]`, so maybe theres a deeper bug somewhere?

[the remark this answers is not in the document]
Comment (KuuCi) [reply llmfoundry/models/llm_embed/modeling_llm_embed.py:215]: also note that `step_size` should just be `pos_step_size` in original code. Perhaps this is a bug? but admittedly, the loss curve looks **so** much more unhelpful
https://github.com/mosaicml/llm-foundry/blob/main/llmfoundry/models/llm_embed/modeling_llm_embed.py#L160

[the remark this answers is not in the document]
Comment (mrdrozdov) [reply llmfoundry/models/llm_embed/modeling_llm_embed.py:218]: @KuuCi can you confirm that your input file has negatives?

[the remark this answers is not in the document]
Comment (KuuCi) [reply llmfoundry/models/llm_embed/modeling_llm_embed.py:216]: Yup had negatives. Met with @jacobfulano, am pretty convinced that `self.step_size = input_shape[1]` now. 

self.step_size = input_shape[1]

I have this debug:
embedding-ft-no-pos-neg-sSVxbf which is off of:
https://github.com/mosaicml/llm-foundry/blob/f9550b4d777f344863854bdfd0479c4851999ecb/llmfoundry/models/llm_embed/modeling_llm_embed.py

I see a very reasonable curve with `gather_in_batch_negatives = False` and upping the batch size
Screenshot 2024-11-08 at 12 21 39 PM

File: llmfoundry/models/llm_embed/modeling_llm_embed.py [hunk 2]
```diff
@@ -209,23 +208,38 @@ def construct_model(self):
...
+        if input_shape[1] > 2:
+            # We have hard negatives, batch shape is [batch, sample of query+positive passage+negative passages, tokens].
+            self.step_size = input_shape[1]
+            log.info(
+                f'Detected hard negatives, updated step_size to {self.step_size}',
+            )
+            self.gather_in_batch_negatives = False
```
Comment (mrdrozdov) [inline llmfoundry/models/llm_embed/modeling_llm_embed.py:223]: Oh, hmm.

Ideally, we could do in-batch negatives AND hard negatives at the same time.

IDK if the code supported this before, and suggest we simply retain the existing behavior. In either case, in-batch negatives is something that should be configurable IMO.

Comment (KuuCi) [reply llmfoundry/models/llm_embed/modeling_llm_embed.py]: Thats reasonable

File: llmfoundry/models/llm_embed/modeling_llm_embed.py [hunk 3]
```diff
@@ -219,13 +233,13 @@ def format_queries_batch(
         Here ``n`` is the step size, which represents the number of hard
         negatives per passage.
         """
+        assert self.step_size
         queries = {}
+        indices = []
```
Comment (mrdrozdov) [inline llmfoundry/models/llm_embed/modeling_llm_embed.py:238]: The default behavior is a bit confusing. Can you add a comment?

File: llmfoundry/models/llm_embed/modeling_llm_embed.py [hunk 4]
```diff
@@ -237,24 +251,22 @@ def format_passages_batch(
...
-
-        # Index on a variable step size
-        index = 0
+        num_blocks = batch['input_ids'].size(0) // self.step_size
+        index = torch.arange(
+            1,
+            num_blocks * self.step_size + 1,
+            device=last_hidden_state.device,
+        ).view(num_blocks, self.step_size)
+        index = index[:, :self.step_size - 1].reshape(-1)
```
Comment (mrdrozdov) [inline llmfoundry/models/llm_embed/modeling_llm_embed.py:262]: Good change!

**Review by mrdrozdov — CHANGES_REQUESTED** [review]
This looks mostly good to me!

A few requests for integration tests:

* Confirm expected behavior when training data has no hard negatives.
* Confirm when using hard negatives = 1.
* Confirm when using hard negatives = 3 (or some number besides 1).

Maybe you've already done this? In which case I will switch to approve after you address my minor comments.

File: llmfoundry/models/llm_embed/modeling_llm_embed.py (hunk 3 above)
Comment (KuuCi) [reply llmfoundry/models/llm_embed/modeling_llm_embed.py:238]: This is just for pyright.

Comment (mrdrozdov) [reply llmfoundry/models/llm_embed/modeling_llm_embed.py:238]: I see. It would be nice if this was tidier I guess.

What about something like:

```
# outside the loop
indices = list(range(0, last_hidden_state.size(0), self.step_size))

for key in batch:
  queries[key] = batch[key][indices, :]
return queries, last_hidden_state[indices, :, :]
```

**Comment by KuuCi** [comment]
Tested with the following runs:

0 negatives: `embedding-ft-no-pos-neg-0-7jyk87`
image

1 negative: `embedding-ft-no-pos-neg-1-zAeAzg`
image

20 negatives: `embedding-ft-no-pos-neg-20-EicOP8`
image

All looks good

**Review by mrdrozdov — APPROVED** [review]
:shipit:

**Review by milocress — APPROVED** [review]
